File: go/futures/mexcfutures/http_client.go
package mexcfutures

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

// RawResponse holds full HTTP response information for a REST call.
// It keeps the original body so callers can log/inspect the exact server payload.
type RawResponse struct {
	StatusCode int    // HTTP status code
	Body       []byte // Raw response body bytes
	Text       string // Raw response body as string (helper)
}

// DecodeJSON decodes the raw body into v. It does not modify the response itself.
func (r *RawResponse) DecodeJSON(v any) error {
	if r == nil {
		return fmt.Errorf("nil RawResponse")
	}
	if len(r.Body) == 0 {
		return nil
	}
	return json.Unmarshal(r.Body, v)
}

// HTTPClient performs MEXC Futures REST requests (public and private).
type HTTPClient struct {
	cfg    Config
	client *http.Client
}

// NewHTTPClient builds an HTTP client with the given config (timeout from config).
func NewHTTPClient(cfg Config) *HTTPClient {
	timeout := cfg.Timeout
	if timeout <= 0 {
		timeout = DefaultTimeout
	}
	return &HTTPClient{
		cfg: cfg,
		client: &http.Client{
			Timeout: timeout,
		},
	}
}

// hasAuth reports whether API key and secret are configured (for private endpoints).
func (c *HTTPClient) hasAuth() bool {
	return c.cfg.APIKey != "" && c.cfg.APISecret != ""
}

// PublicRequest sends a request without signature. params is used as query for GET/DELETE or ignored for POST (use body in that case).
// For POST, pass body as second argument or as params; here we use a single params for both query and body for simplicity; method decides.
func (c *HTTPClient) PublicRequest(ctx context.Context, method, path string, params map[string]interface{}) (*RawResponse, error) {
	return c.doRequest(ctx, method, path, params, nil, false)
}

// PrivateRequest sends a signed request. For GET/DELETE params are query; for POST params are JSON body (or pass body as raw map/array).
func (c *HTTPClient) PrivateRequest(ctx context.Context, method, path string, params interface{}) (*RawResponse, error) {
	return c.doRequest(ctx, method, path, params, nil, true)
}

func (c *HTTPClient) doRequest(ctx context.Context, method, path string, params interface{}, bodyBytes []byte, private bool) (*RawResponse, error) {
	baseURL := strings.TrimSuffix(c.cfg.BaseURL, "/")
	fullPath := baseURL + path
	var reqBody io.Reader
	var query string

	switch method {
	case http.MethodGet, http.MethodDelete:
		m := toParamsMap(params)
		filtered := FilterParams(m)
		query = buildEncodedQuery(filtered)
		if query != "" {
			fullPath = fullPath + "?" + query
		}
	case http.MethodPost:
		if bodyBytes != nil {
			reqBody = bytes.NewReader(bodyBytes)
		} else {
			payload, err := buildPostBody(params)
			if err != nil {
				return nil, err
			}
			reqBody = bytes.NewReader(payload)
		}
	default:
		return nil, fmt.Errorf("unsupported method: %s", method)
	}

	req, err := http.NewRequestWithContext(ctx, method, fullPath, reqBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.cfg.UserAgent != "" {
		req.Header.Set("User-Agent", c.cfg.UserAgent)
	}

	if private {
		requestTime := strconv.FormatInt(time.Now().UnixMilli(), 10)
		// Sign with same params used for query (GET/DELETE) or body (POST).
		sig, err := Sign(c.cfg.APIKey, c.cfg.APISecret, requestTime, method, params)
		if err != nil {
			return nil, err
		}
		req.Header.Set("ApiKey", c.cfg.APIKey)
		req.Header.Set("Request-Time", requestTime)
		req.Header.Set("Signature", sig)
		if c.cfg.RecvWindow > 0 {
			req.Header.Set("Recv-Window", strconv.FormatInt(c.cfg.RecvWindow, 10))
		}
		if c.cfg.Debug {
			debugLogRequest(method, fullPath, req.Header, reqBody, true)
		}
	} else if c.cfg.Debug {
		debugLogRequest(method, fullPath, req.Header, reqBody, false)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	out := &RawResponse{
		StatusCode: resp.StatusCode,
		Body:       raw,
		Text:       string(raw),
	}

	var decoded map[string]interface{}
	if len(raw) > 0 {
		_ = json.Unmarshal(raw, &decoded)
	}

	if resp.StatusCode >= 400 {
		apiErr := &APIError{
			HTTPStatus: resp.StatusCode,
			RawBody:    raw,
		}
		if decoded != nil {
			if c, ok := decoded["code"].(float64); ok {
				apiErr.Code = int(c)
			}
			if m, ok := decoded["message"].(string); ok {
				apiErr.Message = m
			}
		}
		return out, apiErr
	}
	// Optional: treat success=false as error even with 200
	if decoded != nil {
		if s, ok := decoded["success"].(bool); ok && !s {
			apiErr := &APIError{HTTPStatus: resp.StatusCode, RawBody: raw}
			if c, ok := decoded["code"].(float64); ok {
				apiErr.Code = int(c)
			}
			if m, ok := decoded["message"].(string); ok {
				apiErr.Message = m
			}
			return out, apiErr
		}
	}
	return out, nil
}

func toParamsMap(v interface{}) map[string]interface{} {
	if v == nil {
		return nil
	}
	m, ok := v.(map[string]interface{})
	if ok {
		return m
	}
	return nil
}

func buildEncodedQuery(params map[string]interface{}) string {
	if len(params) == 0 {
		return ""
	}
	keys := make([]string, 0, len(params))
	for k := range params {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	vals := make(url.Values)
	for _, k := range keys {
		vals.Set(k, queryValueString(params[k]))
	}
	return vals.Encode()
}

// queryValueString converts a param value to string for URL query (same semantics as util for consistency).
func queryValueString(v interface{}) string {
	if v == nil {
		return ""
	}
	switch x := v.(type) {
	case string:
		return x
	case float64:
		if x == float64(int64(x)) {
			return strconv.FormatInt(int64(x), 10)
		}
		return strconv.FormatFloat(x, 'f', -1, 64)
	case int:
		return strconv.Itoa(x)
	case int64:
		return strconv.FormatInt(x, 10)
	case bool:
		if x {
			return "true"
		}
		return "false"
	default:
		data, _ := json.Marshal(v)
		return string(data)
	}
}

func buildPostBody(params interface{}) ([]byte, error) {
	if params == nil {
		return []byte("{}"), nil
	}
	switch v := params.(type) {
	case map[string]interface{}:
		filtered := FilterParams(v)
		return json.Marshal(filtered)
	case []interface{}:
		return json.Marshal(v)
	default:
		return json.Marshal(params)
	}
}

// debugLogRequest logs method and URL only; never logs APISecret or Signature.
func debugLogRequest(method, fullURL string, headers http.Header, body io.Reader, redactSignature bool) {
	log.Printf("[mexcfutures] %s %s", method, fullURL)
	if redactSignature && headers.Get("Signature") != "" {
		log.Printf("[mexcfutures] Signature: ***")
	}
	_ = headers
	_ = body
}

File: node.js/futures/httpClient.js
/**
 * MEXC Futures HTTP client.
 * Uses Futures OpenAPI signature rules.
 */

const axios = require('axios')
const { sign, filterParams } = require('./signer')

const DEFAULT_BASE_URL = 'https://api.mexc.com'
const DEFAULT_RECV_WINDOW = 10000

/**
 * Convert params to URL query string (for GET/DELETE).
 * Raw key=value used for signing (no URL encode); actual request URL is encoded.
 */
function paramsToQueryString (params) {
  const filtered = filterParams(params)
  const keys = Object.keys(filtered).sort()
  return keys
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(String(filtered[k]))}`)
    .join('&')
}

/**
 * Send request.
 * @param {Object} options
 * @param {string} options.baseURL
 * @param {string} options.method - GET | POST | DELETE
 * @param {string} options.path - Path, e.g. /api/v1/contract/ping
 * @param {Object} options.params - Business params
 * @param {string} [options.apiKey]
 * @param {string} [options.apiSecret]
 * @param {number} [options.recvWindow] - Optional, milliseconds
 * @param {number} [options.timeout] - Optional, request timeout milliseconds
 * @param {boolean} [options.unwrapData=false] - If true, return only data field on success; failure still returns full body
 * @param {Object} [options.headers] - Extra headers, merged with default (e.g. timezone-login)
 * @param {Object} [options.logger]
 * @returns {Promise}
 */
function request (options) {
  const {
    baseURL = DEFAULT_BASE_URL,
    method = 'GET',
    path,
    params = {},
    apiKey,
    apiSecret,
    recvWindow = DEFAULT_RECV_WINDOW,
    timeout,
    unwrapData = false,
    headers: extraHeaders,
    logger
  } = options

  const isPrivate = !!(apiKey && apiSecret)
  const requestTime = String(Date.now())

  let url = path
  let body = null
  const headers = {
    'Content-Type': 'application/json'
  }
  if (extraHeaders && typeof extraHeaders === 'object') {
    Object.assign(headers, extraHeaders)
  }

  if (method === 'GET' || method === 'DELETE') {
    const qs = paramsToQueryString(params)
    if (qs) {
      url = `${path}${path.includes('?') ? '&' : '?'}${qs}`
    }
  } else if (method === 'POST') {
    body = Array.isArray(params) ? params : filterParams(params)
  }

  if (isPrivate) {
    const signature = sign({
      accessKey: apiKey,
      apiSecret,
      requestTime,
      method,
      params: method === 'POST' ? body : params
    })

    headers['ApiKey'] = apiKey
    headers['Request-Time'] = requestTime
    headers['Signature'] = signature
    if (recvWindow != null) {
      headers['Recv-Window'] = String(recvWindow)
    }
  }

  const config = {
    baseURL,
    method,
    url,
    headers,
    data: body,
    validateStatus: () => true
  }
  if (timeout != null && timeout > 0) {
    config.timeout = timeout
  }

  if (logger && typeof logger.debug === 'function') {
    logger.debug('[Futures HTTP]', method, url, body ? '(body)' : '')
  }

  return axios(config).then(res => {
    const body = res.data
    if (logger && res.status >= 400 && typeof logger.error === 'function') {
      logger.error('[Futures HTTP]', res.status, body)
    }
    if (unwrapData && body && body.success === true && 'data' in body) {
      return body.data
    }
    return body != null ? body : res.data
  })
}

/**
 * Public request (no signature).
 */
function publicRequest (method, path, params = {}, baseURL = DEFAULT_BASE_URL, logger) {
  return request({
    baseURL,
    method,
    path,
    params,
    logger
  })
}

/**
 * Private request (with signature).
 */
function privateRequest (method, path, params = {}, options = {}) {
  const { apiKey, apiSecret, baseURL = DEFAULT_BASE_URL, recvWindow, timeout, logger } = options
  return request({
    baseURL,
    method,
    path,
    params,
    apiKey,
    apiSecret,
    recvWindow,
    timeout,
    logger
  })
}

module.exports = {
  request,
  publicRequest,
  privateRequest,
  DEFAULT_BASE_URL,
  DEFAULT_RECV_WINDOW,
  paramsToQueryString
}
